### PR TITLE
Add tunnel mode config and egress gateway config params

### PR DESCRIPTION
**Pull request by blzhao-0**
This PR adds a tunnel mode arg to the init script of cilium devices so
that 'mode' and 'tunnel mode' is represented in separate params. It 
also adds the config options for enabling egress gateway and makes
sure that tunnel is enabled when egress gateway is specified.

This change is part the egress NAT feature, Ref: #13575.

Signed-off-by: Bolun Zhao <blzhao@google.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Thanks for contributing!

### CODE REVIEW

File: bpf/init.sh
Comment (joestringer) [inline bpf/init.sh:485]: Could you double-check the rest of the `$MODE` usage in the file? It looks like there are places where it could be `ipip`, which I presume should be handled via the `${TUNNEL_MODE}` option as well.

**Review by joestringer — CHANGES_REQUESTED** [review]
Hi @MasterZ40 , thanks for the PR.

The code change itself looks mostly straightforward, but I don't understand why. Would you mind providing a brief motivation in the commit message / PR description?

Please also review the rest of the PR description, everything below "Please ensure your pull request adheres to the following guidelines" is for you to complete or remove if it makes no sense. In particular the release note needs to be either set, or just removed (it's fine to remove in this case since the PR title describes the changes sufficiently).

One more specific change request below.

File: bpf/init.sh
Comment (anfernee) [reply bpf/init.sh:485]: Yes. `$MODE` could be `ipip`. It's not used for tunnel mode, but it should be covered. @MasterZ40 can you check it out and fix it?

Comment (blzhao-0) [reply bpf/init.sh:485]: Thanks for the comments. 
Sorry if i did not fully understood, but it seems like `ipip` mode is handled separately in this [file.](https://github.com/cilium/cilium/blob/72949b4b49d43e6dbb4714929222acb688ce6dd2/bpf/init.sh#L441) @borkmann could you comment?

File: bpf/init.sh
Comment (pchaigno) [inline bpf/init.sh]: AFAIU, this comment is for `NATIVE_DEVS`, not for `TUNNEL_MODE`. We might need a similar comment for `TUNNEL_MODE` though.

File: bpf/init.sh
Comment (pchaigno) [reply bpf/init.sh:485]: @MasterZ40 The condition here should check for VXLAN or GENEVE. We don't want to create a `cilium_ipip` device for IPIP.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: bpf/init.sh
Comment (borkmann) [reply bpf/init.sh:485]: @MasterZ40 the ipip is handled separately, mostly given we need separate devices for v4/v6 which is not the case for vxlan + geneve. Also the kernel device name handling is more hacky for the ipip ones which ignores passed name when it needs to pull in the driver.

For testing ipip I typically use: `./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true --datapath-mode=lb-only --bpf-lb-algorithm=maglev --bpf-lb-maglev-table-size=2039 --bpf-lb-mode=dsr --bpf-lb-acceleration=native --devices=enp2s0np0 --bpf-lb-dsr-dispatch=ipip --disable-envoy-version-check=true`

File: bpf/init.sh
Comment (blzhao-0) [reply bpf/init.sh]: Fixed. Thanks for the feedback!

File: bpf/init.sh
Comment (blzhao-0) [reply bpf/init.sh:485]: @borkmann thanks for the clarification.  @pchaigno I think in this case as Daniel said the ipip devices are handled separately as datapath tunnel devices. And this change is not really touching that separation. The ipip devices are handled here: https://github.com/cilium/cilium/blob/72949b4b49d43e6dbb4714929222acb688ce6dd2/bpf/init.sh#L441 

I might be wrong here, but let me know if this makes sense. Thx!

Comment (pchaigno) [reply bpf/init.sh:485]: Ah, I thought IPIP would be one of the tunnel modes, but it's actually one of the modes. Ok :+1:

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Comment by anfernee** [comment]
Do we need this commit: https://github.com/cilium/cilium/pull/14723/commits/f4ecd644bc2fa4439c937098778d3843ccab13d1 ?

**Comment by pchaigno** [comment]
> Do we need this commit: [f4ecd64](https://github.com/cilium/cilium/commit/f4ecd644bc2fa4439c937098778d3843ccab13d1) ?

It's needed for the egress gateway, no? Might as well include it now.

**Comment by anfernee** [comment]
> > Do we need this commit: [f4ecd64](https://github.com/cilium/cilium/commit/f4ecd644bc2fa4439c937098778d3843ccab13d1) ?
> 
> It's needed for the egress gateway, no? Might as well include it now.

Yes. I'd like to hold this one for now for this refactor.

**Comment by MasterZ40** [comment]
> > > Do we need this commit: [f4ecd64](https://github.com/cilium/cilium/commit/f4ecd644bc2fa4439c937098778d3843ccab13d1) ?
> > 
> > 
> > It's needed for the egress gateway, no? Might as well include it now.
> 
> Yes. I'd like to hold this one for now for this refactor.

Synced with @anfernee we are holding this pr to only the refactor.

**Comment by pchaigno** [comment]
test-me-please

**Comment by MasterZ40** [comment]
> > > > Do we need this commit: [f4ecd64](https://github.com/cilium/cilium/commit/f4ecd644bc2fa4439c937098778d3843ccab13d1) ?
> > > 
> > > 
> > > It's needed for the egress gateway, no? Might as well include it now.
> > 
> > 
> > Yes. I'd like to hold this one for now for this refactor.
> 
> Synced with @anfernee we are holding this pr to only the refactor.

Per further discussion I added back in the config for egress gateway commits with clearer descriptions of these options and their relationships.

**Comment by joestringer** [comment]
@MasterZ40 will you update the PR description specifically to remove the "fixes" and release note sections since they don't make sense? The release note piece in particular will break release scripting if you do not either update or remove it.

**Comment by MasterZ40** [comment]
> @MasterZ40 will you update the PR description specifically to remove the "fixes" and release note sections since they don't make sense? The release note piece in particular will break release scripting if you do not either update or remove it.

Updated PR description. Thanks!

**Review by joestringer — APPROVED** [review]
The changes themselves look fine to me. You'll need to take into account MTU for mixed mode as well, otherwise endpoints may end up transmitting packets that cannot fit in the tunnel (I believe this is currently switching on tunnel mode enablement).

Then there's just the question of whether to merge PRs that contain partially-finished features. There has been some ongoing debate about this. I think previously the PR was just refactoring so it's clear we would just merge it, but with the latest changes to also start including the beginnings of the feature for #13575, it's less clear.

EDIT: I guess on the latter aspect, the [ship has already departed](https://github.com/cilium/cilium/pull/14712). So I think the remaining item here is just the other codeowners.

**Comment by pchaigno** [comment]
test-me-please

**Comment by anfernee** [comment]
ready to merge this one?

**Comment by pchaigno** [comment]
Yes, I think so. Tests are green and we have two reviews. Not all team reviews are covered, but I think the changes are trivial.

**Pull request merged**
